tools/bidder-emulator/main.go

```diff
@@ -0,0 +1,271 @@
+		EnvVars: []string{"TRANSACTOR_LOG_TAGS"},
+		Action: func(ctx *cli.Context, s string) error {
+			for i, p := range strings.Split(s, ",") {
+				if len(strings.Split(p, ":")) != 2 {
+					return fmt.Errorf("invalid log-tags at index %d, expecting <name:value>", i)
+				}
+			}
+			return nil
+		},
+	}
+)
+
+func main() {
+	app := &cli.App{
+		Name:  "bidder-emulator",
+		Usage: "Issue random transactions for L1 chain from specified accounts and bid on them on mev-commit chain",
+		Flags: []cli.Flag{
+			optionKeystorePathPassword,
+			optionL1RPCURL,
+			optionBidderRPCURL,
+			optionDepositAmount,
+			optionBidWorkers,
+			optionLogFmt,
+			optionLogLevel,
+			optionLogTags,
+		},
+		Action: func(c *cli.Context) error {
+			l1RPC, err := ethclient.Dial(c.String(optionL1RPCURL.Name))
+			if err != nil {
+				return err
+			}
+
+			logger, err := util.NewLogger(
+				c.String(optionLogLevel.Name),
+				c.String(optionLogFmt.Name),
+				c.String(optionLogTags.Name),
+				c.App.Writer,
+			)
+			if err != nil {
+				return fmt.Errorf("failed to create logger: %w", err)
+			}
+
+			chainID, err := l1RPC.ChainID(context.Background())
+			if err != nil {
+				return fmt.Errorf("failed getting chain ID: %w", err)
+			}
+
+			txtors := make([]*transactorAccount, 0)
+			for _, kp := range c.StringSlice(optionKeystorePathPassword.Name) {
+				parts := strings.Split(kp, ":")
+				t, err := newTransactorAccount(chainID, path.Dir(parts[0]), parts[1], l1RPC)
+				if err != nil {
+					return err
+				}
+				txtors = append(txtors, t)
+			}
+
+			bidderClient, err := newBidder(c.String(optionBidderRPCURL.Name), c.String(optionDepositAmount.Name))
+			if err != nil {
+				return fmt.Errorf("failed to create bidder client: %w", err)
+			}
+
+			ctx, cancel := signal.NotifyContext(c.Context, syscall.SIGINT, syscall.SIGTERM)
+			defer cancel()
+
+			mtx := multex.New[string]()
+
+			eg, egCtx := errgroup.WithContext(ctx)
+			for i := 0; i < c.Int(optionBidWorkers.Name); i++ {
+				eg.Go(func() error {
+					return bidWorker(egCtx, logger, txtors, bidderClient, l1RPC, mtx)
+				})
+			}
+
+			return eg.Wait()
+		},
+	}
+
+	if err := app.Run(os.Args); err != nil {
+		fmt.Fprintf(os.Stderr, "error: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+func bidWorker(
+	ctx context.Context,
+	logger *slog.Logger,
+	txtors []*transactorAccount,
+	bidderClient *bidder,
+	l1RPC *ethclient.Client,
+	mtx *multex.Multex[string],
+) error {
+	if len(txtors) < 2 {
+		return fmt.Errorf("at least 2 transactor accounts are required")
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		start := time.Now()
+
+		rand.Shuffle(len(txtors), func(i, j int) {
+			txtors[i], txtors[j] = txtors[j], txtors[i]
+		})
+
+		from := txtors[0]
+		to := txtors[1]
+
+		// random amount between 0 and 1_000_000_000_000
+		amount := big.NewInt(rand.Int64N(1_000_000_000_000) + 1)
+
+		mtx.Lock(from.Address().Hex())
+		txn, err := from.SendTransaction(ctx, to.Address(), amount)
```

Comment:
Shouldn't we not need to separately `SendTransaction` if the tx payload is a part of the bid?

Reply:
Basically the problem this emulator solves is that we have 2 different services right now. One is the l1-transaction-emulator which creates the L1 transactions and then the bidder-emulator which picks transactions from the blocks to bid on.

The new emulator basically combines the two. This results in better control when we have to scale up. Each worker will first create a transaction on mock L1 and then once the transaction is mined, it will send a bid for the same transaction. Problem is we dont have block-builders, which is why we have to wait for the txn to be mined to know block number to use in the bid. This will enable us to test the entire flow.